### ISSUE TITLE
ReferenceError: ngDevMode is not defined

### ISSUE THREAD

**Issue by mcalmus** [angular/angular#31595]
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core

### Is this a regression?

 Not sure

### Description

 A clear and concise description of the problem...
If there are compilation errors in a template (while using "ng serve"), I am seeing the error "ReferenceError: ngDevMode is not defined" when running in Firefox (60.6.1esr) or Edge. Chrome shows the actual compilation errors.

## 🔬 Minimal Reproduction

 https://stackblitz.com/...
Create simple app with a component. Introduce compilation errors. Try to run app in browser.

The error (with Angular 8.1.1 is in core.js:13393

## 🔥 Exception or Error
<pre><code>

ReferenceError: ngDevMode is not defined
</code></pre>

## 🌍  Your Environment

Angular Version: 8.1.1 (also 8.1.0)
<pre><code>

</code></pre>

**Anything else relevant?**

**Comment by miroslavkovac** [comment]
I am seeing the same error after upgrading from Angular 6 after a long time. The error is in generated `vendors~main.a7289f63f220e84f1b85.chunk.js` and it looks like couple of `ngDevMode` variables names are wrongly put inside generated JS file:
```
...
ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode, ngDevMode;
var qr = function() {
    return Promise.resolve(null)
}();
function Kr(r) {
...
```

**Comment by mezhik91** [comment]
same issue. @angular/core 8.1.2
Is there any way to fix it?

**Comment by VlasovSergey** [comment]
Same here,
I have this issue in only in production build.
Tried to find the reason. Looks like the issue here
https://github.com/angular/angular/blob/cb848b94102855d1e773892b25375304102cab69/packages/core/src/render3/instructions/lview_debug.ts#L60
I think it should be replaced with 
`export const LViewArray = (typeof ngDevMode !== 'undefined') && createNamedArrayType('LView');`
cause ngDevMode does not exist in prodaction build.

**Comment by mcalmus** [comment]
FWIW: I was not using a production build when I encountered the issue.

**Comment by jfborie** [comment]
Same error with --prod, it looks like this class is gone in 8.2.0-rc.0.

**Comment by CapLonelyFlaw** [comment]
Temporary fixed this issue by declaring ngDevMode inside index.html before all scripts only for prod build
` <% if (htmlWebpackPlugin.options.metadata.ENV === 'production') { %>
    <script>
      var ngDevMode = false;
    </script>
  <% } %>`

**Comment by CapLonelyFlaw** [comment]
> Same here,
> I have this issue in only in production build.
> Tried to find the reason. Looks like the issue here
> https://github.com/angular/angular/blob/cb848b94102855d1e773892b25375304102cab69/packages/core/src/render3/instructions/lview_debug.ts#L60
> 
> 
> I think it should be replaced with
> `export const LViewArray = (typeof ngDevMode !== 'undefined') && createNamedArrayType('LView');`
> cause ngDevMode does not exist in production build, not sure though.

ngDevMode variable used not only at `lview_debug.ts#L60` but everywhere).
Do you use webpack or angular cli?

**Comment by pshurygin** [comment]
I see this error for all my production builds even without compilation errors both with 8.1.3 and 8.2.0-rc.0

**Comment by hoonzis** [comment]
Angular 8.2.0, I am using webpack configuration and running into the same issue when the  '@angular-devkit/build-optimizer/webpack-loader' is part of the bundling process. I suppose this loadeder removes that variable.

Removing the loader from the rules fixes the issue (but results in not-optimized builds).
This can be fixed by injecting the missing variables into the bundle, I have used webpack.DefinePlugin for that:

```
plugins.push(
      new BuildOptimizerWebpackPlugin(),
      new AngularCompilerPlugin({
        tsConfigPath: './tsconfig-aot.json',
        entryModule: './src/app/app.module#AppModule'
      }),
      new webpack.DefinePlugin({
        ngDevMode: false,
        ngI18nClosureMode: false
      })
    )

```

ngDevMode and ngI18nClosureMode are normally both defined in:

https://github.com/angular/angular/blob/78146c1890b472621acf33d778477db8df816043/packages/compiler-cli/src/tooling.ts

**Comment by JonWallsten** [comment]
@alxhub: Not sure this is freq: low anymore. Seems like a lot of us has this issue. I guess it applies to anyone using Webpack/AngularCompilerPlugin

**Comment by JonWallsten** [comment]
@filipesilva: I noticed you added a test case for GLOBAL_DEFS_FOR_TERSER a couple of months ago. I'm trying to figure out when and how these are added into the project. Can't find any references in the project for it.

**Comment by filipesilva** [comment]
@JonWallsten `GLOBAL_DEFS_FOR_TERSER` is meant to be imported by the build system, and sets `ngDevMode` to a hardcoded false.

Does anyone have a precise reproduction for seeing these errors at runtime so I can investigate? When I do `ng serve --prod` on a new application I don't see any `ReferenceError: ngDevMode is not defined` at runtime.

**Comment by JonWallsten** [comment]
@filipesilva Let me see if I can reproduce in my in my Webpack AOT Angular repo.

**Comment by JonWallsten** [comment]
@filipesilva: It's easily reproducible with a minimal repo:
Repo: https://github.com/JonWallsten/angular-bug
Step 1: npm i
Step 2: npm run build
Step 3: ...
Step 4: Profit!

**Comment by jfborie** [comment]
I use too custom web pack loading, I use lazy loading with aot. I look at https://webpack.js.org/configuration/mode/ and took as a sample Mode: development, and change what i wanted. Final result is.
const TerserPlugin = require('terser-webpack-plugin');
const webpack = require('webpack');

module.exports = {
  optimization:
  {
    minimize: true,
    namedModules: true,
    namedChunks: true,
    flagIncludedChunks: false,
    occurrenceOrder: false,
    sideEffects: false,
    usedExports: false,
    concatenateModules: false,
    splitChunks: {
      hidePathInfo: false,
      minSize: 10000,
      maxAsyncRequests: Infinity,
      maxInitialRequests: Infinity,
    },
    removeAvailableModules: false,
    minimizer:
    [
      new TerserPlugin
      (
        {
          cache: false,
          parallel: true,
          terserOptions:
          {
            comments: false,
            mangle: false,
            keep_classnames: true,
            keep_fnames: true
          }
        }
      )
    ]
  },
  plugins: [
    new webpack.NamedModulesPlugin(),
    new webpack.NamedChunksPlugin()
  ]
};

And this work for me.

**Comment by filipesilva** [comment]
@JonWallsten thank you for the repro, I'll try to dig in but it might take a bit to isolate the problem since it's a custom setup.

Is anyone seeing this with Angular CLI too?

**Comment by JonWallsten** [comment]
@filipesilva: I just added "npm run build:prod:debug" in case you need to use Chrome DevTools to debug the code. I would suggest you start looking at webpack.prod.ts to see if the config somehow is the error behind this. For example I have implemented Terser.

> @JonWallsten GLOBAL_DEFS_FOR_TERSER is meant to be imported by the build system, and sets ngDevMode to a hardcoded false.

Like in your build system, when you create the bundles för Angular, or when we build our bundles?
If it's the latter I guess there would be some possible explanation since we build in a custom way.

**Comment by filipesilva** [comment]
Ok, I know what's going on. If you make a new project and run `ng serve --prod`, the served project has no error.

Then if you open `node_modules/@angular-devkit/build-angular/src/angular-cli-files/models/webpack-configs/common.js` and remove these line:
```
                    passes: buildOptions.buildOptimizer ? 3 : 1,
                    // global_defs: angularGlobalDefinitions, // <-- comment this one
```
And run `ng serve --prod` again, you can see this error at runtime:
```
Uncaught ReferenceError: ngDevMode is not defined
    at Module.zUnb (core.js.pre-build-optimizer.js:8325)
    at f (bootstrap:79)
    at Object.0 (events.js:426)
    at f (bootstrap:79)
    at t (bootstrap:45)
    at Array.r [as push] (bootstrap:32)
    at main.78106fc448395a14fbe1.js:1
```

You can work around this problem in your custom webpack setup by importing our provided Terser global definition [like we do in Angular CLI](https://github.com/angular/angular-cli/blob/9478ee9c47fbf835d32f4ae1dcf7c428fb1b7eed/packages/angular_devkit/build_angular/src/angular-cli-files/models/webpack-configs/common.ts#L300-L304).
```
const terserOptions = {
  // ...
  compress: {
    // ...
    global_defs: require('@angular/compiler-cli').GLOBAL_DEFS_FOR_TERSER,
  }
}
```

But I think this is a bug because code in `@angular/core` shouldn't break when optimized without these definitions.

The source line where the error comes from is this one: https://github.com/angular/angular/blob/961d663fbef4d6ea2ebb5a60f38799a273fa71ed/packages/core/src/render3/instructions/lview_debug.ts#L58

There are other similar usages throughout the codebase. It is indeed an error to reference something that isn't defined. We can see this in a chrome console or node repl:
```
kamik@RED-X1C6 MINGW64 /d/sandbox
$ node
> var something = missingGlobalVariable && true
Thrown:
ReferenceError: missingGlobalVariable is not defined
> global.missingGlobalVariable = false
false
> var something = missingGlobalVariable && true
undefined
> console.log(something)
false
undefined
```

I think we need to change the code in`@angular/core` to not have errors when `ngDevMode` is not defined.

/cc @kara @mhevery @IgorMinar

**Comment by JonWallsten** [comment]
@filipesilva Wow, that timing though! Take a look at my edit I did 18 seconds after your post ;)

**Comment by mhevery** [comment]
@filipesilva wrote
> I think we need to change the code in`@angular/core` to not have errors when `ngDevMode` is not defined.

But we have such code here: https://github.com/angular/angular/blob/939529ce5dcc7a526a67d45723fbd8ede83d7750/packages/core/src/util/ng_dev_mode.ts#L110-L112 So there is something else going on. Did that code go somehow removed?

**Comment by JonWallsten** [comment]
@mhevery: Problem is in these files:
https://github.com/angular/angular/blob/939529ce5dcc7a526a67d45723fbd8ede83d7750/packages/core/src/render3/instructions/lview_debug.ts
https://github.com/angular/angular/blob/7b9891d7cd1744fb5a1f16a4e4f3e46901db41ef/packages/core/src/render3/instructions/shared.ts

**Comment by filipesilva** [comment]
@JonWallsten I think when team members paste permalinks they get converted into those snippets automatically. https://github.com/angular/angular/pull/32079 should address this problem.

**Comment by mcalmus** [comment]
I'm still getting this error using 9.0.0-next.7. The stack trace is slightly different. Now it's coming from initNgDevMode in core.js:1404. This does not happen in Chrome but happens in Firefox, Edge and IE.

**Comment by filipesilva** [comment]
Reopening to investigate.

**Comment by filipesilva** [comment]
I followed the steps in https://github.com/angular/angular/issues/31595#issuecomment-519129090 and was able to see the error again as @mcalmus reported. I had to set `"enableIvy": false` in `tsconfig.json` and use `ng serve --prod --source-map` to get good source maps.

The error came from this line:
https://github.com/angular/angular/blob/04ddcfecd8134e4c57ce95ce35444d1e26b424a3/packages/core/src/core.ts#L42

This is indeed a problem because this line is a toplevel side effect that checks a variable that might not exist at the toplevel. It needs to treated similarly as other toplevel side effects were in https://github.com/angular/angular/pull/32079, preferably in such a way that doesn't have a leftover side effect like it does now in the [side-effects test](https://github.com/angular/angular/blob/04ddcfecd8134e4c57ce95ce35444d1e26b424a3/integration/side-effects/snapshots/core/esm2015.js#L5-L17) (cc @petebacondarwin).

If I replace it `if (false) {` the error goes away so I think it should be ok.

Your report of the 1404 line is worrisome though... It's this done:
```
function initNgDevMode() {
    // The below checks are to ensure that calling `initNgDevMode` multiple times does not
    // reset the counters.
    // If the `ngDevMode` is not an object, then it means we have not created the perf counters
    // yet.
    if (typeof ngDevMode === 'undefined' || ngDevMode) {
        if (typeof ngDevMode !== 'object') {
            ngDevModeResetPerfCounters(); 
        }
        return !!ngDevMode; //    <----- this is 1404
    }
    return false;
}
```

I'm not sure how you get there but it makes sense that it's a problem, because `ngDevMode` is initialized in `ngDevModeResetPerfCounters` only. I think this logic needs to be revisited.

**Comment by mcalmus** [comment]
> I'm not sure how you get there but it makes sense that it's a problem, because `ngDevMode` is initialized in `ngDevModeResetPerfCounters` only. I think this logic needs to be revisited.

That's the right line. Unfortunately the project is large and somewhat complicated, and I've been unable to figure out a more basic project with which to reproduce. I'll keep trying. I hope knowing the line number will at least help.

**Comment by mcalmus** [comment]
Does a new PR need to be created for the additional patching?

**Comment by mhevery** [comment]
@arnehoek could you turn your patch into a PR so that it can be merged?

**Comment by khteh** [comment]
Bump into the same error. Any npm package update for the fix?

**Comment by eng-dibo** [comment]
I think this issue is cause by terser dist/browser/main-*.js file

**Comment by mcalmus** [comment]
I realized the problem I was having was because I had a global variable re-used from some legacy code named "global" declared in global.d.ts.

That said, I think this pull request would eliminate the error even in that scenario:

https://github.com/angular/angular/pull/36055

**Comment by JoostK** [comment]
The fix in #36055 wasn't quite adequate in avoiding this error, the `$localize` check in `@angular/core` is still remaining. I opened #39415 to resolve that one, which will close this issue once it's been merged.
